agent/agent.go

```diff
@@ -431,26 +416,110 @@ func (a *Agent) processNextTask(ctx context.Context, nt *apimodels.NextTaskRespo
+			oomTracker:                jasper.NewOOMTracker(),
+			unsetFunctionVarsDisabled: nt.UnsetFunctionVarsDisabled,
+		}
+	} else {
+		tc = initialTC
+	}
+
+	// If the heartbeat aborts the task immediately, we should report that
+	// the task failed during initial task setup.
+	factory, ok := command.GetCommandFactory("setup.initial")
+	if !ok {
+		return tc, shouldExit, errors.New("setup.initial command is not registered")
+	}
+	tc.setCurrentCommand(factory())
+	a.comm.UpdateLastMessageTime()
+
+	var taskConfig *internal.TaskConfig
+	taskConfig, err = a.makeTaskConfig(setupCtx, tc)
+	if err != nil {
+		tc.logger = client.NewSingleChannelLogHarness("agent.error", a.defaultLogger)
+		return a.handleSetupError(setupCtx, tc, errors.Wrap(err, "making task config"))
+	}
+	tc.setTaskConfig(taskConfig)
+	if err := a.startLogging(agentCtx, tc); err != nil {
+		tc.logger = client.NewSingleChannelLogHarness("agent.error", a.defaultLogger)
+		return a.handleSetupError(setupCtx, tc, errors.Wrap(err, "setting up logger producer"))
+	}
+
+	if !tc.ranSetupGroup {
+		tc.taskDirectory, err = a.createTaskDirectory(tc)
+		if err != nil {
+			tc.logger.Execution().Error(errors.Wrap(err, "creating task directory"))
+			return a.handleSetupError(setupCtx, tc, errors.Wrap(err, "creating task directory"))
+		}
+	}
+	tc.taskConfig.WorkDir = tc.taskDirectory
+	tc.taskConfig.Expansions.Put("workdir", tc.taskConfig.WorkDir)
+
+	factory, ok = command.GetCommandFactory("setup.initial")
+	if !ok {
+		tc.logger.Execution().Error("Marking task as system-failed because setup.initial command is not registered.")
+		return tc, shouldExit, errors.New("setup.initial command is not registered")
+	}
+	tc.setCurrentCommand(factory())
+	a.comm.UpdateLastMessageTime()
+
+	tc.logger.Task().Infof("Task logger initialized (agent version '%s' from Evergreen build revision '%s').", evergreen.AgentVersion, evergreen.BuildRevision)
+	tc.logger.Execution().Info("Execution logger initialized.")
+	tc.logger.System().Info("System logger initialized.")
+
+	if err := setupCtx.Err(); err != nil {
+		grip.Error(err)
+		tc.logger.Execution().Infof("Stopping task execution: %s", err)
+		shouldExit, err := a.handleTaskResponse(setupCtx, tc, evergreen.TaskSystemFailed, err.Error())
+		return tc, shouldExit, err
+	}
+
+	hostname, err := os.Hostname()
+	tc.logger.Execution().Info(errors.Wrap(err, "getting hostname"))
+	if hostname != "" {
+		tc.logger.Execution().Infof("Hostname is '%s'.", hostname)
+	}
+	tc.logger.Task().Infof("Starting task '%s', execution %d.", tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution)
+
+	return tc, shouldExit, nil
 }
 
+func (a *Agent) handleSetupError(ctx context.Context, tc *taskContext, err error) (*taskContext, bool, error) {
+	catcher := grip.NewBasicCatcher()
+	grip.Error(err)
+	catcher.Add(err)
+	grip.Infof("Task complete: '%s'.", tc.task.ID)
+	shouldExit, err := a.handleTaskResponse(ctx, tc, evergreen.TaskSystemFailed, err.Error())
+	catcher.Add(err)
```

Comment:
I would `catcher.Wrap(err, "handling task response")` (same with the error caught above) so that we can distinguish between this error and the error caught above this. Otherwise, it'll be weird because the resolved catcher errors, when logged, will look like one error happened (the setup error), then it caused a completely random set of errors happened (but it didn't because it's a separate error).